---
title: "BayeScan input files"
author: "Grace Smith-Vidaurre"
date: "September 14, 2020"
output: html_document
---

```{r setup, eval = TRUE, echo = FALSE}

knitr::opts_knit$set(root.dir = "/home/owner/Desktop/GitHub_repos/origins-selection")

```

Purpose: Subset the Structure file of the HWE and missing data filtered merged SNPs to make versions representing all pairwise population comparisons to be executed in BayeScan. The files generated in this script will be in Structure format, and will be converted to BayeScan formt using a script to execute PGDSpider.

```{r echo = TRUE, eval = TRUE, message = FALSE}

rm(list = ls())

X <- c("tidyverse", "pbapply", "data.table", "adegenet", "openxlsx")
invisible(lapply(X, library, character.only = TRUE))

# Path to the metadata spreadsheet that will be updated
xls_path <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/DATA/Metadata_Barcodes"

# Path to Stacks output, including the HWE filtered SNPs in Structure format 
res_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/stacks"

# Path where BayeScan files will be written
out_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/BayeScan_mergedSNPs"

gpath <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/GRAPHICS"

seed <- 401

```

Read in metadata.
```{r echo = TRUE, eval = TRUE}

meta_dats <- read.xlsx(file.path(xls_path, "Mymon_RADlibraries_2015_2019_CombinedMetadata.xlsx"))
glimpse(meta_dats)

```

Read in the HWE and missing data filtered merged SNPs with duplicated positive controls dropped.
```{r echo = TRUE, eval = TRUE}

struc <- read.table(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters_noPosControlDups.str"), skip = 1)
# glimpse(struc)
dim(struc) # 346 rows (173 individuals), 563 columns
head(struc[, 1]) # Sample names contained in the first column
head(struc[, 2]) # Population ID (single population) contained in second column
names(struc)

# Get the column headers

# Read the first line of metadata, these are the column names
col_nms <- readLines(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters_noPosControlDups.str"), n = 1)
col_nms

# Split out the column headers into a vector, one element per header
mrkrs <- strsplit(strsplit(col_nms, split = "\t")[[1]][3], split = "[[:space:]]")[[1]]
head(mrkrs)
length(mrkrs)

# Get rid of the space that precedes the first locus column header
mrkrs <- mrkrs[-which(mrkrs == "")]
head(mrkrs)
length(mrkrs)

# Add the individual and population column headers that precede the locus column names
col_nms <- c("indiv", "pop", mrkrs)
head(col_nms)
length(col_nms)

# Add back the updated column names
# Looks good
names(struc) <- col_nms
glimpse(struc[, 1:10])

```

# Write out files for population contrasts

These are the general population contrasts I'm considering. Some of these will entail multiple files/comparisons (e.g. 14-18):

1) All invasive sites versus all native sites
2) All invasive sites versus all Uruguay sites
3) All invasive sites versus all Argentina sites
4) All invasive sites versus all northern Argentina sites
5) All invasive sites versus all southern Argentina sites
6) All U.S. sites versus all Uruguay sites
7) All U.S. sites versus all Argentina sites
8) All U.S. sites versus all northern Argentina sites
9) All U.S. sites versus all southern Argentina sites
10) All Spain sites versus all Uruguay sites
11) All Spain sites versus all Argentina sites
12) All Spain sites versus all northern Argentina sites
13) All Spain sites versus all southern Argentina sites
14) All United States versus all Spain sites
15) Each invasive site versus all native sites (9 comparisons)
16) Each invasive site versus all Uruguay sites (9 comparisons)
17) Each invasive site versus all Argentina sites (9 comparisons)
18) Each invasive site versus all northern Argentina sites (9 comparisons)
19) Each invasive site versus all southern native sites (9 comparisons)
20) Randomizations


1) All invasive sites versus all native sites. Change the population column encode invasive versus native.
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  dplyr::select(-c(Population))

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
#   slice(1:50) %>%
#   dplyr::select(indiv, pop) %>%
#   View()

# Write a new file
file_nm <- "BS_01_AllInvasiveVersusNative.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good
# Use :set list in Vim to see invisible characters (tabs)

```

2) All invasive sites versus all Uruguay sites.
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Spain", "Uruguay")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) # 173 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
#   dplyr::select(indiv, pop) %>%
#   View()

# Write a new file
file_nm <- "BS_02_AllInvasiveVersusUruguay.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

3) All invasive sites versus all Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Spain", "Argentina")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 169 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
#   dplyr::select(indiv, pop) %>%
#   View()

# Write a new file
file_nm <- "BS_03_AllInvasiveVersusArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

4) All invasive sites versus all northern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Spain") | Country == "Argentina" & !Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 144 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
#   dplyr::select(indiv, pop) %>%
#   View()

# Write a new file
file_nm <- "BS_04_AllInvasiveVersusNorthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

5) All invasive sites versus all southern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Spain") | Country == "Argentina" & Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 152 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_05_AllInvasiveVersusSouthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

6) All U.S. sites versus all Uruguay sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Uruguay")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 133 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_06_UnitedStatesVersusUruguay.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

7) All U.S. sites versus all Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("United States", "Argentina")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 128 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_07_UnitedStatesVersusArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

8) All U.S. sites versus all northern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country == "United States" | Country == "Argentina" & !Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 103 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_08_UnitedStatesVersusNorthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

9) All U.S. sites versus all southern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country == "United States" | Country == "Argentina" & Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 111 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_09_UnitedStatesVersusSouthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

10) All Spain sites versus all Uruguay sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("Spain", "Uruguay")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 88 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_10_SpainVersusUruguay.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

11) All Spain sites versus all Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country %in% c("Spain", "Argentina")) %>%
  dplyr::select(-c(Population, Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 83 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_11_SpainVersusArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

12) All Spain sites versus all northern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country == "Spain" | Country == "Argentina" & !Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 58 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_12_SpainVersusNorthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

13) All Spain sites versus all southern Argentina sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Population, Country, Site_Code),
    by = c("indiv" = "Sample_Name")
  ) %>%
  dplyr::mutate(
    pop = factor(Population, levels = c("Native", "Invasive")),
    pop = as.numeric(pop)
  ) %>%
  filter(Country == "Spain" | Country == "Argentina" & Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
  dplyr::select(-c(Population, Country, Site_Code))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 66 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_13_SpainVersusSouthArgentina.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

14) All United States versus all Spain sites
```{r echo = TRUE, eval = TRUE}

struc_tmp <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Country),
    by = c("indiv" = "Sample_Name")
  ) %>%
  filter(Country %in% c("United States", "Spain")) %>%
  dplyr::mutate(
    pop = factor(Country, levels = c("United States", "Spain")),
    pop = as.numeric(pop)
  ) %>%
  dplyr::select(-c(Country))

dim(struc_tmp) 
nrow(struc_tmp)/2 # 127 individuals

# Looks good
glimpse(struc_tmp[, 1:10])
# struc_tmp %>%
  # dplyr::select(indiv, pop) %>%
  # View()

# Write a new file
file_nm <- "BS_14_UnitedStatesVersusSpain.str"
file.remove(file.path(out_path, file_nm))

header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

cat(header, file = file.path(out_path, file_nm), append = TRUE)

invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
  cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
}))

# Opened the file in Vim to doublecheck structure, looks good

```

15) Each invasive site versus all native sites (9 comparisons). Exclude SEVI because this was just one individual.
```{r echo = TRUE, eval = TRUE}

inv_sites <- meta_dats %>% 
  filter(Population == "Invasive") %>%
  filter(Site_Code != "SEVI") %>%
  pull(Site_Code) %>%
  unique() %>%
  as.character()

inv_sites
length(inv_sites)

# Iterate over invasive range sites to create one file each
# i <- 1
invisible(pblapply(1:length(inv_sites), function(i){
  
  inv_struc_tmp <- struc %>% 
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Site_Code),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Site_Code == inv_sites[i]) %>%
    dplyr::mutate(
      pop = 2
    ) %>%
    dplyr::select(-c(Site_Code))
  
  # glimpse(inv_struc_tmp[, 1:10])
  # inv_struc_tmp[, 1:10] %>%
    # View()
  
  nat_struc_tmp <- struc %>%
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Population),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Population == "Native") %>%
    dplyr::mutate(
      pop = 1
    ) %>%
    dplyr::select(-c(Population))
  
   # glimpse(nat_struc_tmp[, 1:10])
   # nat_struc_tmp[, 1:10] %>%
    # View()
   
   # Combine native and invasive range data
   struc_tmp <- inv_struc_tmp %>%
     bind_rows(nat_struc_tmp)

  # dim(struc_tmp) 
  # nrow(struc_tmp)/2
  # struc_tmp[, 1:5] %>%
  #   View()
   
  # Write a new file
  file_nm <- paste("BS_15-", i, "_", inv_sites[i], "_", "VersusAllNative.str", sep = "")
  # file_nm
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened each file in Vim to doublecheck structure, these look good

```

16) Each invasive site versus all Uruguay sites
```{r echo = TRUE, eval = TRUE}

inv_sites <- meta_dats %>% 
  filter(Population == "Invasive") %>%
  filter(Site_Code != "SEVI") %>%
  pull(Site_Code) %>%
  unique() %>%
  as.character()

inv_sites
length(inv_sites)

# Iterate over invasive range sites to create one file each
# i <- 1
invisible(pblapply(1:length(inv_sites), function(i){
  
  inv_struc_tmp <- struc %>% 
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Site_Code),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Site_Code == inv_sites[i]) %>%
    dplyr::mutate(
      pop = 2
    ) %>%
    dplyr::select(-c(Site_Code))
  
  # glimpse(inv_struc_tmp[, 1:10])
  # inv_struc_tmp[, 1:10] %>%
    # View()
  
  nat_struc_tmp <- struc %>%
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Country),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Country == "Uruguay") %>%
    dplyr::mutate(
      pop = 1
    ) %>%
    dplyr::select(-c(Country))
  
   # glimpse(nat_struc_tmp[, 1:10])
   # nat_struc_tmp[, 1:10] %>%
    # View()
   
   # Combine native and invasive range data
   struc_tmp <- inv_struc_tmp %>%
     bind_rows(nat_struc_tmp)

  # dim(struc_tmp) 
  # nrow(struc_tmp)/2
  # struc_tmp[, 1:5] %>%
  #   View()
   
  # Write a new file
  file_nm <- paste("BS_16-", i, "_", inv_sites[i], "_", "VersusUruguay.str", sep = "")
  # file_nm
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened each file in Vim to doublecheck structure, these look good

```

17) Each invasive site versus all Argentina sites
```{r echo = TRUE, eval = TRUE}

inv_sites <- meta_dats %>% 
  filter(Population == "Invasive") %>%
  filter(Site_Code != "SEVI") %>%
  pull(Site_Code) %>%
  unique() %>%
  as.character()

inv_sites
length(inv_sites)

# Iterate over invasive range sites to create one file each
# i <- 1
invisible(pblapply(1:length(inv_sites), function(i){
  
  inv_struc_tmp <- struc %>% 
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Site_Code),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Site_Code == inv_sites[i]) %>%
    dplyr::mutate(
      pop = 2
    ) %>%
    dplyr::select(-c(Site_Code))
  
  # glimpse(inv_struc_tmp[, 1:10])
  # inv_struc_tmp[, 1:10] %>%
    # View()
  
  nat_struc_tmp <- struc %>%
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Country),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Country == "Argentina") %>%
    dplyr::mutate(
      pop = 1
    ) %>%
    dplyr::select(-c(Country))
  
   # glimpse(nat_struc_tmp[, 1:10])
   # nat_struc_tmp[, 1:10] %>%
    # View()
   
   # Combine native and invasive range data
   struc_tmp <- inv_struc_tmp %>%
     bind_rows(nat_struc_tmp)

  # dim(struc_tmp) 
  # nrow(struc_tmp)/2
  # struc_tmp[, 1:5] %>%
  #   View()
   
  # Write a new file
  file_nm <- paste("BS_17-", i, "_", inv_sites[i], "_", "VersusArgentina.str", sep = "")
  # file_nm
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened each file in Vim to doublecheck structure, these look good

```

18) Each invasive site versus all northern Argentina sites
```{r echo = TRUE, eval = TRUE}

inv_sites <- meta_dats %>% 
  filter(Population == "Invasive") %>%
  filter(Site_Code != "SEVI") %>%
  pull(Site_Code) %>%
  unique() %>%
  as.character()

inv_sites
length(inv_sites)

# Iterate over invasive range sites to create one file each
# i <- 1
invisible(pblapply(1:length(inv_sites), function(i){
  
  inv_struc_tmp <- struc %>% 
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Site_Code),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Site_Code == inv_sites[i]) %>%
    dplyr::mutate(
      pop = 2
    ) %>%
    dplyr::select(-c(Site_Code))
  
  # glimpse(inv_struc_tmp[, 1:10])
  # inv_struc_tmp[, 1:10] %>%
    # View()
  
   nat_struc_tmp <- struc %>% 
     inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Country, Site_Code),
        by = c("indiv" = "Sample_Name")
      ) %>%
      dplyr::mutate(
        pop = 1
      ) %>%
      filter(Country == "Argentina" & !Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
      dplyr::select(-c(Country, Site_Code))
  
   # glimpse(nat_struc_tmp[, 1:10])
   # nat_struc_tmp[, 1:10] %>%
    # View()
   
   # Combine native and invasive range data
   struc_tmp <- inv_struc_tmp %>%
     bind_rows(nat_struc_tmp)

  # dim(struc_tmp) 
  # nrow(struc_tmp)/2
  # struc_tmp[, 1:5] %>%
  #   View()
   
  # Write a new file
  file_nm <- paste("BS_18-", i, "_", inv_sites[i], "_", "VersusNorthArgentina.str", sep = "")
  # file_nm
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened each file in Vim to doublecheck structure, these look good

```

19) Each invasive site versus all southern Argentina sites
```{r echo = TRUE, eval = TRUE}

inv_sites <- meta_dats %>% 
  filter(Population == "Invasive") %>%
  filter(Site_Code != "SEVI") %>%
  pull(Site_Code) %>%
  unique() %>%
  as.character()

inv_sites
length(inv_sites)

# Iterate over invasive range sites to create one file each
# i <- 1
invisible(pblapply(1:length(inv_sites), function(i){
  
  inv_struc_tmp <- struc %>% 
    inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Site_Code),
      by = c("indiv" = "Sample_Name")
    ) %>%
    filter(Site_Code == inv_sites[i]) %>%
    dplyr::mutate(
      pop = 2
    ) %>%
    dplyr::select(-c(Site_Code))
  
  # glimpse(inv_struc_tmp[, 1:10])
  # inv_struc_tmp[, 1:10] %>%
    # View()
  
   nat_struc_tmp <- struc %>% 
     inner_join(
      meta_dats %>%
        dplyr::select(Sample_Name, Country, Site_Code),
        by = c("indiv" = "Sample_Name")
      ) %>%
      dplyr::mutate(
        pop = 1
      ) %>%
      filter(Country == "Argentina" & Site_Code %in% c("ALGA", "BAIR", "LURO")) %>%
      dplyr::select(-c(Country, Site_Code))
  
   # glimpse(nat_struc_tmp[, 1:10])
   # nat_struc_tmp[, 1:10] %>%
    # View()
   
   # Combine native and invasive range data
   struc_tmp <- inv_struc_tmp %>%
     bind_rows(nat_struc_tmp)

  # dim(struc_tmp) 
  # nrow(struc_tmp)/2
  # struc_tmp[, 1:5] %>%
  #   View()
   
  # Write a new file
  file_nm <- paste("BS_19-", i, "_", inv_sites[i], "_", "VersusSouthArgentina.str", sep = "")
  # file_nm
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened each file in Vim to doublecheck structure, these look good

```

20) Randomizations to assess whether outlier loci detected were just as likely to have been detected by chance. As in Lemay & Russello 2015, randomly place all individuals into one of two groups, do this five times.
```{r echo = TRUE, eval = TRUE}

iter <- 5

# Randomly select 5 seeds for the random sampling below, such that the randomization can be replicated
set.seed(seed)
seeds <- sample(100, 5, replace = FALSE)
seeds

# i <- 1
invisible(pblapply(1:iter, function(i){
  
  set.seed(seeds[i])
  
  indivs_rsamp <- struc %>%
    pull(indiv) %>%
    unique() %>%
    as.character() %>%
    sample(length(.)/2, replace = FALSE)
  
  struc_tmp <- struc %>% 
    dplyr::mutate(
      indiv = as.character(indiv),
      pop = ifelse(indiv %in% indivs_rsamp, 1, 2)
    )
  
  # glimpse(struc_tmp[, 1:5])
  # struc_tmp[, 1:5] %>%
    # View()

  # Write a new file
  file_nm <- paste("BS_20-", i, "_RandomContrast.str", sep = "")
  # file_nm
  
  file.remove(file.path(out_path, file_nm))

  header <- c("\t","\t", paste(paste(mrkrs, collapse = " "), "\n", sep = ""))

  cat(header, file = file.path(out_path, file_nm), append = TRUE)

  invisible(pbsapply(1:nrow(struc_tmp), function(i){
  
    cat(c(struc_tmp$indiv[i], "\t", struc_tmp$pop[i], "\t", paste(paste(struc_tmp[i, -grep("^indiv$|^pop$", names(struc_tmp))], collapse = " "), "\n", sep = "")), file = file.path(out_path, file_nm), append = TRUE)
  
  }))
  
}))

# Opened the files in Vim to doublecheck structure, these look good

# 64 Bayescan files, including 5 random contrasts as validation
length(list.files(path = file.path(out_path), pattern = "BS_*"))

```

# Convert Structure to BayeScan format

Reworking code I'd written previously for analysis of the single-end libraries. Although BayeScan performs faster when monomorphic loci are dropped, I did not drop these here. I believe PGDSpider doesn't either, and for the merged SNPs, we're working with a small set of loci anyways (hundreds and not thousands).

Bayescan format: 
Header with [loci]=NumberofLoci
Blank line
[populations]=Totalpopulations
Blank line
[pop]=1
1(two spaces)40(two spaces)2(three spaces)0(two spaces)40
One line per locus as above, until done for the first population
Blank line
[pop]=2
Locus lines as above

Per locus, the numbers are as follows: Consecutive numeric identifier for the given locus, total number of genes for the marker in the population (if diploid, then twice the number of individuals, can be modified to account for missing data per locus), The number of alleles for the given locus (must be the same across populations, and here will always be 2), all subsequent numbers are the number of observations for each allele in the population for the given locus, and must sum to the number of genes in the population.

In the example line above, this is the 1st locus with 40 total genes in the population and 2 alleles. The first allele was observed 0 times, and the second was observed 40 times.

Make a function to return the locus information in the same format as above per population. The function takes a Structure file as input, and returns locus information in BayeScan format.
```{r echo = TRUE, eval = FALSE}

pop_locus_info <- function(X, mrkrs, NA_char){ 
  
  pops <- length(unique(X$pop))
  
  # Iterate over populations
  pop_list <- pblapply(1:pops, function(i){
    
    # Subset the Structure input file by the given population
    X_tmp <- X[X$pop == i, ]
    
    # Iterate over loci per population
    tmp_loci <- unlist(pblapply(1:length(mrkrs), function(j){
      
      # Subset the data for the given population to obtain just the given locus
      tmp <- X_tmp[, grep(paste("^", paste(mrkrs[j], "$", sep = ""), sep = ""), colnames(X_tmp))]
  
      # Account for missing data at the given locus to avoid counting missing data as an allele
      if(NA_char %in% tmp){
        n.alleles <- length(unique(tmp)) - 1
      } else {
        n.alleles <- length(unique(tmp))
      }
  
      # Account for missing data at the given locus to make the number of genes in the population reflect any missing data across individuals
      missing <- length(which(tmp == NA_char))
      
      # For diploids, the number of genes should be twice the number of individuals that have data
      n.genes <- length(tmp) - missing
  
      # Insert message about loci with more than 2 alleles 
      # All loci have 2 alleles, so any loci with 3 means that the NA_char was not specified correctly
      if(n.alleles == 3){
        stop("Are you using the correct NA_char to indicate missing data?")
      }
  
      # Printing the number of diploiad observations per allele
      # Account for loci with a single allele
      if(n.alleles == 1){
    
        tmp2 <- unique(tmp)[grep(NA_char, unique(tmp), invert = TRUE)]
    
        first.allele <- length(which(tmp == tmp2[1]))
        second.allele <- c(0)
        n.obs.alleles <- paste(c(first.allele, second.allele), collapse = "  ")
    
      # Stop if sum of observed alleles not equal to total genes at this locus
      if(first.allele != n.genes){
        stop(paste("j = ", j, ", n.alleles == 1: Observed alleles at this locus are not equal to total expected
           genes (number diploid individuals * 2)"))
      }
    
    # If the given locus has 2 alleles, account for order and numeric coding of SNPs when writing out the locus info
    # PGDSpider always starts with the lower allele first (possible numeric values of 1, 2, 3, 4 in Structure genotypes)
    } else if(n.alleles == 2){
    
      tmp2 <- unique(tmp)[grep(NA_char, unique(tmp), invert = TRUE)]
    
      # If the second allele is a larger number than the first (e.g. 4 versus 2)
      if(tmp2[2] > tmp2[1]){
        first.allele <- length(which(tmp == tmp2[1]))
        second.allele <- length(which(tmp == tmp2[2]))
        n.obs.alleles <- paste(c(first.allele, second.allele), collapse = "  ")
      
        # Stop if sum of observed alleles not equal to total genes at this locus
        if(first.allele + second.allele != n.genes){
          stop(paste("j = ", j, ", n.alleles == 2: Observed alleles at this locus are not equal to total expected
           genes (number diploid individuals * 2)"))
        }
      }
    
      # If the first allele is a larger number than the second
      if(tmp2[1] > tmp2[2]){
        first.allele <- length(which(tmp == tmp2[2]))
        second.allele <- length(which(tmp == tmp2[1]))
        n.obs.alleles <- paste(c(first.allele, second.allele), collapse = "  ")
      
        # Stop if sum of observed alleles not equal to total genes at this locus
        if(first.allele + second.allele != n.genes){
            stop(paste("j = ", j, ", Observed alleles at this locus are not equal to total expected
           genes (number diploid individuals * 2)"))
          }
        }
      }
      
      # Concatenate the information for the given locus in the given population
      locus_info <- paste(
        paste(
          c(# If the locus index is less than 3 characters, pad with whitespace to 3 characters
            ifelse(nchar(j) < 3, str_pad(j, width = 3, side = "left"), j),
            # Add 2 leading whitespace characters to n.genes and n.alleles
            str_pad(n.genes, width = nchar(n.genes) + 2, side = "left"),
            str_pad(n.alleles, width = nchar(n.alleles) + 2, side = "left"),
            # Add 3 leading whitespace characters to n.obs.alleles
            str_pad(n.obs.alleles, width = nchar(n.obs.alleles) + 3, side = "left")
            ), 
          collapse = ""), 
        "\n", sep = "")
      
      # locus_info
      
      return(locus_info)
       
    }))
    
    # str(tmp_loci)
    # head(tmp_loci)
    return(tmp_loci)
    
  })

  # str(pop_list)
  return(pop_list)

}

```

Make another function that will run the function over all Structure files generated above. This loop will print BayeScan headers and locus information per population to a new .txt file.
```{r echo = TRUE, eval = FALSE}

Structure2Bayescan <- function(file_nms, n.pops, fpath){
  
  invisible(pblapply(1:length(file_nms), function(z){
  
    # Read in the genotypes from the given Structure file
    struc <- read.table(file.path(fpath, file_nms[z]), skip = 1)
    # glimpse(struc)
    # dim(struc)

    # Get the column headers
    # Read the first line of metadata, these are the column names
    col_nms <- readLines(file.path(fpath, file_nms[z]), n = 1)
    # col_nms

    # Split out the column headers into a vector, one element per header
    mrkrs <- strsplit(strsplit(col_nms, split = "\t")[[1]][3], split = "[[:space:]]")[[1]]
    # head(mrkrs)
    # length(mrkrs)

    # Get rid of the space that precedes the first locus column header
    mrkrs <- mrkrs[-which(mrkrs == "")]
    # head(mrkrs)
    # length(mrkrs)

    # Add the individual and population column headers that precede the locus column names
    col_nms <- c("indiv", "pop", mrkrs)
    # head(col_nms)
    # length(col_nms)

    # Add back the updated column names
    # Looks good
    names(struc) <- col_nms
    # glimpse(struc[, 1:10])

    # Create the overall BayeScan header with the number of loci and populations (whoch will always be 2 per file), and the first population header
    first_header <- list(
      c(paste("[loci]=", length(mrkrs), "\n\n", "[populations]=", n.pops, "\n", sep = "")), 
      c(paste("[pop]=", 1, "\n", sep = ""))
    )
      
    # Get the locus info lines per population
    # X is the Structure object with indiv and pop column headers
    pop_locus_list <- pop_locus_info(X = struc, mrkrs = mrkrs, NA_char = -9999)

    # Create the header for the second population
    second_pop_header <- paste("[pop]=", 2, "\n", sep = "")
  
    # Initialize the BayeScan file name, change just the extension  
    new_file <- file.path(fpath, paste(gsub(".str$", "", file_nms[z]), ".txt", sep = ""))
  
    # Remove any previous version of the given file if it exists
    if(file.exists(new_file)) file.remove(new_file)
      
    # Write out the headers and locus info per population to this new file
    cat(c(
      unlist(first_header[[1]]), 
      "\n",
      unlist(first_header[[2]]), 
      unlist(pop_locus_list[[1]]),
      "\n",
      second_pop_header, 
      unlist(pop_locus_list[[2]])), 
      file = new_file, 
      append = TRUE,
      sep = ""
    )
  
  }))
  
}


# Testing
# file_nms <- "BS_01_AllInvasiveVersusNative.str"
# z <- 1
# fpath <- file.path(out_path)
# n.pops <- 2

fpath <- out_path
file_nms <- list.files(path = fpath, pattern = "^BS_.*.str$")
file_nms

# Running the function for all Structure files generated above
Structure2Bayescan(file_nms = file_nms, n.pops = 2, fpath = fpath)

# Opened 6 files in Vim to doublecheck BayeScan structure, these looked great
list.files(fpath, pattern = ".txt$")

```

Next, need to upload these files to Discovery along with previous Bayescan code, update the Bayescan script, and start a run on Discovery, troubleshoot as needed.
