---
title: "BLASTing outliers"
author: "Grace Smith-Vidaurre"
date: "October 15, 2020"
output: html_document
---

```{r setup, eval = TRUE, echo = FALSE}

knitr::opts_knit$set(root.dir = "/home/owner/Desktop/GitHub_repos/origins-selection")

```

Purpose: Parsing blastn results to run blastx, also parsing blastx results. For outliers and extended sequences of the budgie genome.

```{r echo = TRUE, eval = TRUE, message = FALSE}

rm(list = ls())

X <- c("tidyverse", "pbapply", "data.table", "adegenet", "openxlsx", "XML", "vcfR")
invisible(lapply(X, library, character.only = TRUE))

# Path to the metadata spreadsheet
xls_path <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/DATA/Metadata_Barcodes"

# Path to Stacks output, including the HWE filtered SNPs in Structure format 
res_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/stacks"

# Path where BLAST files will be read and written
path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/BLAST"
 
# Path where population maps written
map_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/info"

gpath <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/GRAPHICS"

seed <- 401
cores <- parallel::detectCores() - 2

```

Read in metadata.
```{r echo = TRUE, eval = TRUE}

meta_dats <- read.xlsx(file.path(xls_path, "Mymon_RADlibraries_2015_2019_CombinedMetadata.xlsx"))
glimpse(meta_dats)

```

Read in the XML file from blastn to the budgie genome.
```{r echo = TRUE, eval = TRUE}

# Read in the XML as an XML
xml_res <- xmlParse(file.path(path, "outlier_RADtags_blastn2Meund.xml"))

# Convert to a list
blastn_res <- xmlToList(xml_res)
glimpse(blastn_res)
length(blastn_res)

blastn_res[[8]] # Parameters used in the call to blastn
blastn_res[[9]] # The last element contains the list of hits
length(blastn_res[[9]]) # One list per outlier locus
str(blastn_res[[9]][[1]])
length(blastn_res[[9]][[1]])

# Iterate over the list of hits and write out results of interest to a data frame
# In doing so, assess the list of multiple hits per outlier locus, and only retain multiple hits if e-values are less than an order of magnitude from the top hit
hit_list <- blastn_res[[9]]

# Do any loci have more than one hit? I think they all have one hit, as expected given the way I parameterized the blastn
sapply(blastn_res[[9]], function(X){
  length(X$`Iteration_hits`)
})

i <- 1
blastn_df <- rbindlist(pblapply(1:length(hit_list), function(i){
  
  # cat(paste("i = ", i, "\n"))
  
  tmp_list <- hit_list[[i]]
  # str(tmp_list)
  # names(tmp_list)
  
  # FASTA header for the outlier locus
  outlier_fasta_headr <- tmp_list$`Iteration_query-def`
  
  # Only proceed with extracting metadata if hits found
  # When no hits are found, there is a last element called "Iteration_message"
  if(!any(grepl("Iteration_message", names(tmp_list)))){
    
    # List of hits 
    # str(tmp_list$`Iteration_hits`)
  
    # Hit ID
    hit_id <- tmp_list$`Iteration_hits`$Hit$Hit_id
  
    # Hit accession number
    hit_acc_num <- tmp_list$`Iteration_hits`$Hit$Hit_accession
  
    # Metdata for the given hit, including accession number, assembly name and chromosome number
    metad <- tmp_list$`Iteration_hits`$Hit$Hit_def
    # metad 
  
    tmp_meta <- strsplit(metad, split = " ")[[1]]
  
    # Chromosome accession number
    chrom_acc_num <- tmp_meta[1]
  
    # Species
    spp <- paste(tmp_meta[2], tmp_meta[3], sep = "_")
  
    # If there is a chromosome number, return this
    # Chromosome number
    if(grepl("chromosome", metad)){
      chrom_num_or_annotation <- paste(tmp_meta[6], gsub(",", "", tmp_meta[7]), sep = "-")
    # Otherwise, return the given annotation (some predicted proteins for instance)
    } else{
      chrom_num_or_annotation <- metad
    }
    
    # Hit length, e.g. the given scaffold or assembled chromosome length
    hit_len <- as.numeric(tmp_list$Iteration_hits$Hit$Hit_len)
    
    # Score for high-scoring segment pair alignment with no gaps
    Hsp_score <- tmp_list$`Iteration_hits`$Hit$Hit_hsps$Hsp$Hsp_score
  
    # E-value for the given hit
    evalue <- tmp_list$`Iteration_hits`$Hit$Hit_hsps$Hsp$Hsp_evalue
  
    # Coordinates for the given hit
    hit_start <- as.numeric(tmp_list$`Iteration_hits`$Hit$Hit_hsps$Hsp$`Hsp_hit-from`)
    hit_end <- as.numeric(tmp_list$`Iteration_hits`$Hit$Hit_hsps$Hsp$`Hsp_hit-to`)
  
    # If the hit end coordinate is larger than the start coordinate, then I'm assuming this is the forward or positive strand
    strand <- ifelse(hit_start < hit_end, "+", "-")
    
    tmp_df <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
      dplyr::mutate(
        hit_id = hit_id,
        hit_acc_num = hit_acc_num,
        chrom_acc_num = chrom_acc_num,
        spp = spp,
        hit_len = hit_len,
        chrom_num_or_annotation = chrom_num_or_annotation,
        Hsp_score = Hsp_score,
        evalue = evalue,
        hit_start = hit_start,
        hit_end = hit_end,
        strand = strand 
      )
   
    # If no hits, just return the locus info
  } else {
    
    tmp_df <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
      dplyr::mutate(
        hit_id = NA,
        hit_acc_num = NA,
        chrom_acc_num = NA,
        spp = NA,
        hit_len = NA,
        chrom_num_or_annotation = NA,
        Hsp_score = NA,
        evalue = NA,
        hit_start = NA,
        hit_end = NA,
        strand = NA
      )
   
  }
  
  return(tmp_df)

  
}))

glimpse(blastn_df)
                       
```

Check out the hits.
```{r echo = TRUE, eval = TRUE}

# How many loci had no hits?
blastn_df %>%
  filter(is.na(hit_id)) %>%
  nrow()
  
# These 19 loci had no hits
blastn_df %>%
  filter(is.na(hit_id)) %>%
  pull(outlier_fasta_headr) %>%
  as.character()

# Among the 222 loci that did have hits, which hits had more annotation beyond chromosome numbers?
blastn_df %>%
  filter(!is.na(hit_id)) %>%
  filter(!grepl("chromosome", chrom_num_or_annotation)) %>%
  group_by(chrom_num_or_annotation) %>%
  dplyr::summarise(
    n = length(hit_id)
  )

# Which proteins are represented?
# 10 predicted proteins, 2 hits to unplaced genomic scaffolds
blastn_df %>%
  filter(!is.na(hit_id)) %>%
  filter(!grepl("chromosome", chrom_num_or_annotation)) %>%
  pull(chrom_num_or_annotation)

# Among the 222 loci that did have hits, which hits had annotation of just chromosome numbers?
blastn_df %>%
  filter(!is.na(hit_id)) %>%
  filter(grepl("chromosome", chrom_num_or_annotation)) %>%
  group_by(chrom_num_or_annotation) %>%
  dplyr::summarise(
    n = length(hit_id)
  )


```

How to proceed from here? It probably makes sense to extend genomic coordinates in either direction...but I'm not sure of how to do this exactly when there are multiple chromosomes. But keep in mind that this blastn was done with the full set of chromosomes that was indexed by samtools faidx. Therefore, the coordinates returned by blastn are from the indexed genome across chromsomes, and samtools faidx should be able to pull out the right coordinates after they've been extended in either direction

Decided to extend genomic coordinates 10kb in either direction. Use Hussein et al. 2020 as a reference to back this up, they used 10kb sliding windows to detect selective sweeps in a rapid avian radiation.

Here, modifying code I'd written previously for analysis with the single-end libraries. Write out a text file that will be fed to samtools faidx. I performed this step for the 222 outlier loci that had hits to the indexed budgie genome.
```{r echo = TRUE, eval = TRUE}

blastn_hits_df <- blastn_df %>%
  filter(!is.na(hit_id)) %>%
  droplevels()
glimpse(blastn_hits_df)

# Extract a 10kb flanking region
flank <- 10000

# Create a new file to hold the lines for samtools faidx that will be used to extract sequences with extended coordinates around original hits
file_nm <- "extended_coordinates4samtools.txt"

# Remove previous versions
file.remove(file.path(path, file_nm))

# Iterate over hits
# i <- 1 # positive strand
# i <- 5 # negative strand
# i <- 212 # positive strand, small genome segment hit
extended_coords <- rbindlist(pblapply(1:nrow(blastn_hits_df), function(i){
  
  # Forward or positive strand is 5'-3', convention is to read from L to R, or positive numeric direction, e.g., start position is smaller than the end position
  # Therefore, the reverse or negative strand will have a start position greater than end position
  # This is already encoded in the column "strand"
  
  # Extend genomic coordinates differently depending on whether the hit is on the positive or negative strand
  # Also, make sure to extend 10kb in either direction if the given hit is far enough away from the start and end of the given genome segment
  if(blastn_hits_df$strand[i] == "+"){
    
    # Find the distance in basepairs (bp), between the hit start and end, and the start and end of the given genome segment
    # Here, the start of the given genome segment is assumed to be 0, and the end is the length of that genome segment (in column "hit_len")
    diff_start <- blastn_hits_df$hit_start[i] - 0
    diff_end <- blastn_hits_df$hit_len[i] - blastn_hits_df$hit_end[i]
      
    # If the distance of the hit coordinates from the start and end of the given genome segment allow for extracting 10kb in either direction, then do so
    if(all(diff_start & diff_end >= flank)){
        
      flank_used <- (flank)
      extended_hit_start <- blastn_hits_df$hit_start[i] - flank_used
      extended_hit_end <- blastn_hits_df$hit_end[i] + flank_used
      
    # If either the start or end difference in coordinates is less than the flanking region, then take the whole genome segment
    } else if(diff_start < flank | diff_end < flank){
        
      flank_used <- "took genome segment"
      extended_hit_start <- 0
      extended_hit_end <- blastn_hits_df$hit_len[i]
      
    }
    
    # Create lines that can be used as input to extract sequences using samtools faidx
    seqs <- paste(blastn_hits_df$chrom_acc_num[i], paste(extended_hit_start, extended_hit_end, sep = "-"), sep = ":")
    
  # Change the logic for the negative strand. Here, the hit_start must be within 10kb of the end of the genome segment, and the hit_end within 10kb of the start of the genome segment
  } else if(blastn_hits_df$strand[i] == "-"){
    
    # Find the distance in basepairs (bp), between the hit start and end, and the start and end of the given genome segment
    # Here, the start of the given genome segment is assumed to be 0, and the end is the length of that genome segment (in column "hit_len")
    diff_start <- blastn_hits_df$hit_len[i] - blastn_hits_df$hit_start[i]
    diff_end <- blastn_hits_df$hit_end[i] - 0
      
    # If the distance of the hit coordinates from the start and end of the given genome segment allow for extracting 10kb in either direction, then do so
    # Note the reversal of the operations compared to the positive strand
    if(all(diff_start & diff_end >= flank)){
        
      extended_hit_start <- blastn_hits_df$hit_start[i] + flank
      extended_hit_end <- blastn_hits_df$hit_end[i] - flank
      flank_used <- (flank)
        
    # If either the start or end difference in coordinates is less than the flanking region, then take the whole genome segment
    } else if(diff_start < flank | diff_end < flank){
        
      flank_used <- "took genome segment"
      extended_hit_start <- blastn_hits_df$hit_len[i]
      extended_hit_end <- 0
      
    }
    
    # Create lines that can be used as input to extract sequences using samtools faidx
    # Here, reversing the start and end for samtools, and will need to check this carefully
    seqs <- paste(blastn_hits_df$chrom_acc_num[i], paste(extended_hit_end, extended_hit_start, sep = "-"), sep = ":")
    
  }

  # Print lines for samtools faidx to the same file as all other lines
  
  # Initialize file name
  tmp_nm <- file.path(path, file_nm)

  # Add two lines per individual, each represents an allele
  cat(paste(seqs, "\n", sep = ""), file = tmp_nm, sep = "", append = TRUE)
    
  # In addition to writing out files for samtools faidx, also return a data frame specifying new coordinates and the size of the flanking region used
  tmp_df <- data.frame(
      locus_id = blastn_hits_df$outlier_fasta_headr[i],
      hit_id = blastn_hits_df$hit_id[i],
      orig_start = blastn_hits_df$hit_start[i],
      orig_end = blastn_hits_df$hit_end[i],
      new_start = extended_hit_start,
      new_end = extended_hit_end,
      strand = blastn_hits_df$strand[i],
      flank_size_used = flank_used
    )
  # class(tmp_df)
  
  return(tmp_df)
  
}))

glimpse(extended_coords)

# I checked a handful of extended coordinates for positive and negative strand hits, looks good
extended_coords %>%
  View()

# None of the new extended coordinates are negative
extended_coords %>%
  filter(new_start < 0 | new_end < 0)

# Sizes of genome segments used when the segments were small?
extended_coords %>%
  filter(flank_size_used == "took genome segment")

# These genome segments ranged from 6048 to 10867bp for positive strands
extended_coords %>%
  filter(flank_size_used == "took genome segment") %>%
  filter(!new_start == 0) %>%
  pull(new_start) %>%
  range()

# And 780 to 10641bp for negative strands
extended_coords %>%
  filter(flank_size_used == "took genome segment") %>%
  filter(!new_end == 0) %>%
  pull(new_end) %>%
  range()

# I opened the .txt file created, it has 222 lines as expected, and separators look good

```

After generating the extended coordinates, use faidx to pull out new FASTA sequences from the budgie genome, then blastx these sequences to the NCBI nr database.

As the extended sequences and the FASTA sequences of the outliers themselves run on Dsicovery, I'm working with blastx results with just the FASTA sequences of outlier loci that I blasted against the budgie genome on NCBI online.
```{r echo = TRUE, eval = TRUE}

# Read in the XML as an XML
xml_res <- xmlParse(file.path(path, "SVE81SK9014-Alignment.xml"))

# Convert to a list
blastx_res <- xmlToList(xml_res)
glimpse(blastx_res)
length(blastx_res)

blastx_res[[8]] # Parameters used in the call to blastx
blastx_res[[9]] # The last element contains the list of hits
length(blastx_res[[9]]) # One list per outlier locus
str(blastx_res[[9]][[1]])
length(blastx_res[[9]][[1]])

# Iterate over the list of hits and write out results of interest to a data frame
# In doing so, assess the list of multiple hits per outlier locus, and only retain multiple hits if e-values are less than an order of magnitude from the top hit
hit_list <- blastx_res[[9]]

# Do any loci have more than one hit? Yes
which(sapply(blastx_res[[9]], function(X){
  length(X$`Iteration_hits`)
}) > 1)

# i <- 1 # no hits
# i <- 16 # hits present
# z <- 1
blastx_df <- rbindlist(pblapply(1:length(hit_list), function(i){
  
  # cat(paste("i = ", i, "\n"))
  
  tmp_list <- hit_list[[i]]
  # str(tmp_list)
  # names(tmp_list)
  
  # FASTA header for the outlier locus
  outlier_fasta_headr <- tmp_list$`Iteration_query-def`
  
  # Only proceed with extracting metadata if hits found
  # Here, when no hits are found, the element called "Iteration_hits" has the value "\n"
  if(length(tmp_list$`Iteration_hits`) > 1){
    
    # List of hits
    # str(tmp_list$`Iteration_hits`)
    
    tmp_hit_list <- tmp_list$`Iteration_hits`
    
    # Iterate over hits to extract information
    tmp_df2 <- rbindlist(lapply(1:length(tmp_hit_list), function(z){
      
      tmp_hit_list2 <- tmp_hit_list[[z]]
      
      # Hit number
      hit_num <- tmp_hit_list2$Hit_num
      
      # Hit ID
      hit_id <- tmp_hit_list2$Hit_id
      
      # Hit annotation
      hit_annot <- tmp_hit_list2$Hit_def
  
      # Hit accession number
      hit_acc_num <- tmp_hit_list2$Hit_accession
  
      # E-value for the given hit
      evalue <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$Hsp_evalue)
  
      # Coordinates for the given hit
      hit_start <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$`Hsp_hit-from`)
      hit_end <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$`Hsp_hit-to`)
  
      # If the hit end coordinate is larger than the start coordinate, then I'm assuming this is the forward or positive strand
      strand <- ifelse(hit_start < hit_end, "+", "-")
    
      tmp_df <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
        dplyr::mutate(
          hit_num = hit_num,
          hit_id = hit_id,
          hit_annot = hit_annot,
          evalue = evalue,
          hit_start = hit_start,
          hit_end = hit_end,
          strand = strand 
        )
      
      return(tmp_df)
  
    }))
   
    # If no hits, just return the locus info
  } else {
    
    tmp_df2 <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
        dplyr::mutate(
          hit_num = NA,
          hit_id = NA,
          hit_annot = NA,
          evalue = NA,
          hit_start = NA,
          hit_end = NA,
          strand = NA 
        )
   
  }
  
  return(tmp_df2)

  
}))

glimpse(blastx_df)

```

Which outlier loci had hits to the genome, and what functional regions are represented?
```{r echo = TRUE, eval = TRUE}

# 10 outlier had hits to budgie protein annotations, anywhere from 2 to 10
blastx_df %>%
  filter(!is.na(hit_num)) %>%
  droplevels() %>%
  group_by(outlier_fasta_headr) %>%
  dplyr::summarise(
    n_hits = length(hit_num)
  )

# How many loci had hits with the e-value threshold I'm currently using of 1e-15?
# Five loci, with 1, 2 or 4 hits
blastx_df %>%
  filter(!is.na(hit_num)) %>%
  filter(evalue <= 1e-15) %>%
  droplevels() %>%
  group_by(outlier_fasta_headr) %>%
  dplyr::summarise(
    n_hits = length(hit_num)
  )

# These are the hits I'm interested in reporting for the thesis chapter
# I took the top hit per locus
hitsOfinterest <- blastx_df %>%
  filter(!is.na(hit_num)) %>%
  filter(evalue <= 1e-15) %>%
  droplevels() 

hitsOfinterest %>%
  View()

# Get the unique accession IDs and annotations
hitsOfinterest %>% 
  dplyr::select(hit_id, hit_annot)

# Just the annotations here
# [1] "retinoblastoma-associated protein [Melopsittacus undulatus]"                                                      
#  [2] "N-acetyl-beta-glucosaminyl-glycoprotein 4-beta-N-acetylgalactosaminyltransferase 1 [Melopsittacus undulatus]"     

#  [4] "breast cancer type 1 susceptibility protein isoform X1 [Melopsittacus undulatus]"                                 
                             
#  [6] "toll-like receptor 4 [Melopsittacus undulatus]"                                                                   
#  [7] "neuron navigator 2 isoform X4 [Melopsittacus undulatus]"                             

```

I searched these hit IDs on NCBI to loop up functionality but that wasn't very informative, so I used Google Scholar or UnitProt website > Complete GO annotation

 - retinoblastoma-associated protein: cell proliferation, division (Henley & Dick 2012)
 
  - N-acetyl-beta-glucosaminyl-glycoprotein 4-beta-N-acetylgalactosaminyltransferase 1: From UniProt, transferase activity, not much more info on this
  
  - breast cancer type 1 susceptibility protein isoform X1: ubiquitin-protein ligase involved in DNA repair, may also be a transcriptional regulator
  
  - toll-like receptor 4: I already knew this but confirmed on UniProt: innate immune response 
  
  - neuron navigator 2 isoforms: From UniProt, in humans these are involved with neuronal development, specifically development of sensory organs, also neurogenesis, locomotor behavior, sensory perception of smell and sound, vagus nerve and optic nerve development
  
Now need to figure out which population comparisons had these outliers.

Read in the dataset of pre-processed neutral SNPs from the merged dataset. The number of individuals and loci is documented in BayeScan_post-processing.Rmd. This dataset contains loci in the same order of the indices returned by BayeScan.
```{r echo = TRUE, eval = TRUE}

struc <- read.table(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters_noPosControlDups.str"), skip = 1)
# glimpse(struc)
dim(struc) # 346 rows (173 individuals), 563 columns
head(struc[, 1]) # Sample names contained in the first column
head(struc[, 2]) # Population ID (single population) contained in second column
names(struc)

# Get the column headers

# Read the first line of metadata, these are the column names
col_nms <- readLines(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters_noPosControlDups.str"), n = 1)
col_nms

# Split out the column headers into a vector, one element per header
mrkrs <- strsplit(strsplit(col_nms, split = "\t")[[1]][3], split = "[[:space:]]")[[1]]
head(mrkrs)
length(mrkrs)

# Get rid of the space that precedes the first locus column header
mrkrs <- mrkrs[-which(mrkrs == "")]
head(mrkrs)
length(mrkrs)

# Add the individual and population column headers that precede the locus column names
col_nms <- c("indiv", "pop", mrkrs)
head(col_nms)
length(col_nms)

# Add back the updated column names
# Convert the indiv column to character
# Looks good
names(struc) <- col_nms
struc$indiv <- as.character(struc$indiv)
glimpse(struc[, 1:10])

```

Get outlier indices from BayeScan results.
```{r echo = TRUE, eval = FALSE}

# Path with BayeScan files
bs_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/BayeScan_mergedSNPs"

# Source the BayeScan R function
# I copied this into my external hard drive from BayeScan2.1/R functions/
source(file.path(bs_path, "plot_R.r"))

# Iterate over files and q-values to find the outliers
res_files <- list.files(path = bs_path, pattern = "_fst.txt$")
qvals <- c(0.2, 0.05)

# 64 comparisons, including 5 randomizations
length(res_files)

i <- 1
q <- 1
outliers_df <- rbindlist(pblapply(1:length(res_files), function(i){
  
  tmp_df2 <- rbindlist(pblapply(1:length(qvals), function(q){
    
    comparison <- gsub("_fst.txt", "", res_files[i])
    
    tmp <- plot_bayescan(res = file.path(bs_path, res_files[i]), FDR = qvals[q])
  
    # Numeric indices of loci, in the same order as the original Structure file
    ols <- tmp$outliers
    
    if(length(ols) > 0){
      tmp_df <- data.frame(
        comparison = comparison,
        q_value = qvals[q],
        outlier_indices = ols
      )
    } else {
      tmp_df <- data.frame(
        comparison = comparison,
        q_value = qvals[q],
        outlier_indices = NA
      )
    }

    return(tmp_df)
    
  }))
  
  return(tmp_df2)
  
}))

outliers_df

```

Get all the unique outlier indices.
```{r echo = TRUE, eval = FALSE}

o_inds <- outliers_df %>%
  filter(!is.na(outlier_indices)) %>%
  pull(outlier_indices) %>%
  unique()

mrkrs[o_inds]

# Get the unique FASTA headers from the FASTA header used for blastx
names(hitsOfinterest)

locus_ids <- hitsOfinterest %>%
  pull(outlier_fasta_headr) %>%
  unique() %>%
  as.character()

# i <- 1
blastx_locus_ids <- sapply(1:length(locus_ids), function(i){
  gsub("CLocus_", "", strsplit(locus_ids[i], split = " ")[[1]][1])
})
blastx_locus_ids

# Search for the blastx locus IDs in the locus names used as input for BayeScan
blastx_inds <- grep(paste(paste("^", blastx_locus_ids, "_", sep = ""), collapse = "|"), mrkrs)

# Confirming that these blastx_inds are consistent with the BayeScan indices
mrkrs[blastx_inds]

# Checking, looks good
# blastx_inds %in% o_inds
# mrkrs[o_inds][grep(paste(paste("^", blastx_locus_ids, "_", sep = ""), collapse = "|"), mrkrs[o_inds])]

# Which comparisons had these indices present among outliers, and at what q-values?
outliers_df %>%
  filter(!is.na(outlier_indices)) %>%
  filter(outlier_indices %in% blastx_inds)

# For now, report only those at a q-value of 0.05
outliers_df %>%
  filter(!is.na(outlier_indices)) %>%
  filter(q_value == 0.05) %>%
  filter(outlier_indices %in% blastx_inds)

```

Read in the VCF file for outlier loci to extract the SNP type per RAD-tag, Also need the accesion ID. The total depth info is really useful too, need to report this for neutral loci certainly.
```{r echo = TRUE, eval = FALSE}

tmp_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/stacks/merged/fstats_regions_outlierSNPs"

vcf_outliers <- read.vcfR(file.path(tmp_path, "populations.snps.vcf"))
str(vcf_outliers)

vcf_outliers@meta

# This object contains the Reference and Alternate SNP indo
vcf_outliers@fix

str(vcf_outliers@gt)

# Reference allele per locus
ref <- vcf_outliers@fix[grep(paste(paste("^", blastx_locus_ids, ":", sep = ""), collapse = "|"), vcf_outliers@fix[, "ID"]), "REF"]

# ALternative allele per locus
alt <- vcf_outliers@fix[grep(paste(paste("^", blastx_locus_ids, ":", sep = ""), collapse = "|"), vcf_outliers@fix[, "ID"]), "ALT"]

```

Combine this metadata in a single data frame to make a table for the main manuscript.
```{r echo = TRUE, eval = FALSE}

# Make a data frame with the locus IDs, functionality, and index
mrkrs[blastx_inds]

# Locus IDs and functionality from above:
# 302: retinoblastoma-associated protein
# 395: N-acetyl-beta-glucosaminyl-glycoprotein 4-beta-N-acetylgalactosaminyltransferase 1
# 1117: breast cancer type 1 susceptibility protein isoform X1
# 3617: toll-like receptor 4
# 3988: neuron navigator 2 isoform X4

annot_df <- data.frame(outlier_indices = blastx_inds) %>%
  dplyr::mutate(
    locus_ids = mrkrs[blastx_inds],
    SNP = paste(ref, alt, sep = "/"), 
    acc_id =c("XP_005147720.2", "XP_033917501.1", "XP_030901062.1", "XP_030901749.2", "XP_005150625.2"),
    annot = c("retinoblastoma-associated protein", "N-acetyl-beta-glucosaminyl-glycoprotein 4-beta-N-acetylgalactosaminyltransferase 1", "breast cancer type 1 susceptibility protein isoform X1", "toll-like receptor 4", "neuron navigator 2 isoform X4"),
    evalues = c(2.94267e-23, 1.32595e-18, 2.42231e-17, 3.94778e-24, 1.98665e-23)
  )

annot_df    

# Join with the BayeScan results to get population comparison info
outliers_df %>%
  filter(!is.na(outlier_indices)) %>%
  filter(q_value == 0.05) %>%
  filter(outlier_indices %in% blastx_inds) %>%
  inner_join(
    annot_df,
    by = "outlier_indices"
  ) %>%
  arrange(-desc(outlier_indices)) %>%
  View()

```

Also get and the locus FST values from the full set of summary statistics for all 561 merged SNPs after HWE filtering, which will allow me to calculate the average FST across all these loci as in Lemay and Russello 2015. See notes below, this may well depend on the population comparisons that displayed the given outlier, and may not be possible to calculate without additional Stacks::populations runs (e.g. ILLI versus Uruguay).
```{r echo = TRUE, eval = FALSE}

# Stacks::populations was run with all neutral and outlier loci, individuals were split into regions here...actually, this needs to be rerun with individuals in a single population, to get FST by locus without population comparisons inthe file populations.fst.tsv...actually, that can't be done because FST is pairwise and therefore requires some sort of population comparison in order for tje value to be calculated
# tmp_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/stacks/merged/fstats_regions"

```

# blastx to full nr database

I also got blastx results from blasting outlier RAD-tags to the full nr database. As above, many RAD-tags did not have any hits, but some did, and these were overall very similar to the hits obtained above to the budgie annotations online.
```{r echo = TRUE, eval = TRUE}

# Read in the XML as an XML
xml_res <- xmlParse(file.path(path, "outlier_RADtags_blastx2nr.xml"))

# Convert to a list
blastx_res <- xmlToList(xml_res)
glimpse(blastx_res)
length(blastx_res)

blastx_res[[8]] # Parameters used in the call to blastx
blastx_res[[9]] # The last element contains the list of hits
length(blastx_res[[9]]) # One list per outlier locus
str(blastx_res[[9]][[1]])
length(blastx_res[[9]][[1]])

# Iterate over the list of hits and write out results of interest to a data frame
# In doing so, assess the list of multiple hits per outlier locus, and only retain multiple hits if e-values are less than an order of magnitude from the top hit
hit_list <- blastx_res[[9]]

# Do any loci have more than one hit? Yes
which(sapply(blastx_res[[9]], function(X){
  length(X$`Iteration_hits`)
}) > 1)

i <- 1 # no hits
# i <- 16 # hits present
# z <- 1
blastx_df <- rbindlist(pblapply(1:length(hit_list), function(i){
  
  # cat(paste("i = ", i, "\n"))
  
  tmp_list <- hit_list[[i]]
  # str(tmp_list)
  # names(tmp_list)
  
  # FASTA header for the outlier locus
  outlier_fasta_headr <- tmp_list$`Iteration_query-def`
  
  # Only proceed with extracting metadata if hits found
  # Here, when no hits are found, the element called "Iteration_hits" has the value "\n"
  if(length(tmp_list$`Iteration_hits`) > 1){
    
    # List of hits
    # str(tmp_list$`Iteration_hits`)
    
    tmp_hit_list <- tmp_list$`Iteration_hits`
    
    # Iterate over hits to extract information
    tmp_df2 <- rbindlist(lapply(1:length(tmp_hit_list), function(z){
      
      tmp_hit_list2 <- tmp_hit_list[[z]]
      
      # Hit number
      hit_num <- tmp_hit_list2$Hit_num
      
      # Hit ID
      hit_id <- tmp_hit_list2$Hit_id
      
      # Hit annotation
      hit_annot <- tmp_hit_list2$Hit_def
  
      # Hit accession number
      hit_acc_num <- tmp_hit_list2$Hit_accession
  
      # E-value for the given hit
      evalue <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$Hsp_evalue)
  
      # Coordinates for the given hit
      hit_start <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$`Hsp_hit-from`)
      hit_end <- as.numeric(tmp_hit_list2$Hit_hsps$Hsp$`Hsp_hit-to`)
  
      # If the hit end coordinate is larger than the start coordinate, then I'm assuming this is the forward or positive strand
      strand <- ifelse(hit_start < hit_end, "+", "-")
    
      tmp_df <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
        dplyr::mutate(
          hit_num = hit_num,
          hit_id = hit_id,
          hit_annot = hit_annot,
          evalue = evalue,
          hit_start = hit_start,
          hit_end = hit_end,
          strand = strand 
        )
      
      return(tmp_df)
  
    }))
   
    # If no hits, just return the locus info
  } else {
    
    tmp_df2 <- data.frame(outlier_fasta_headr = outlier_fasta_headr) %>%
        dplyr::mutate(
          hit_num = NA,
          hit_id = NA,
          hit_annot = NA,
          evalue = NA,
          hit_start = NA,
          hit_end = NA,
          strand = NA 
        )
   
  }
  
  return(tmp_df2)

  
}))

glimpse(blastx_df)

```

Which outlier loci had hits to the genome, and what functional regions are represented?
```{r echo = TRUE, eval = TRUE}

# 8 outlier tags had 5 hits each to protein annotations
blastx_df %>%
  filter(!is.na(hit_num)) %>%
  droplevels() %>%
  group_by(outlier_fasta_headr) %>%
  dplyr::summarise(
    n_hits = length(hit_num)
  )

# How many loci had hits with the e-value threshold I'm currently using of 1e-15?
# All 8 loci, all 5 hits except locus 1117, which had a single hit at this evalue threshold 
blastx_df %>%
  filter(!is.na(hit_num)) %>%
  filter(evalue <= 1e-15) %>%
  droplevels() %>%
  group_by(outlier_fasta_headr) %>%
  dplyr::summarise(
    n_hits = length(hit_num)
  )

# These are the hits I'm interested in reporting for the thesis chapter
# I took the top hit per locus
hitsOfinterest <- blastx_df %>%
  filter(!is.na(hit_num)) %>%
  filter(evalue <= 1e-15) %>%
  droplevels() 

# Many of these are very similar to the hits I found earlier, with at least two new ones, and hits to more parrots and other avian species
# Also includes Chromodomain-helicase-DNA-binding protein 9, which may be a chromatin remodeling protein and is thought to activate PPARA (e.g. it's a transciptional coactivator of this protein), which regulates hundreds of other genes related to lipid metabolism, and may also be related to circadian clocks: https://www.uniprot.org/uniprot/Q3L8U1
hitsOfinterest %>%
  View()

# Get the unique accession IDs and annotations
hitsOfinterest %>% 
  dplyr::select(hit_id, hit_annot)

# Next, need to get the same metadata as above for the other outliers 

```
