---
title: "Exploratory visuals of population genetics structure"
author: "Grace Smith-Vidaurre"
date: "September 12, 2020"
output: html_document
---

```{r setup, eval = TRUE, echo = FALSE}

knitr::opts_knit$set(root.dir = "/home/owner/Desktop/GitHub_repos/origins-selection")

```

Purpose: Make exploratory visuals of population genetic structure with DAPC (discriminant analysis of principal components) with the three sets of HWE filtered SNPs.
```{r echo = TRUE, eval = TRUE, message = FALSE}

rm(list = ls())

X <- c("tidyverse", "pbapply", "data.table", "adegenet", "openxlsx")
invisible(lapply(X, library, character.only = TRUE))

# Path to the metadata spreadsheet that will be updated
xls_path <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/DATA/Metadata_Barcodes"

# Path to Stacks output, including the HWE filtered SNPs in Structure format 
res_path <- "/media/owner/MYIOPSITTA/R/Origins_Selection/stacks"

gpath <- "/home/owner/Desktop/MANUSCRIPTS/Origins_Selection/GRAPHICS"

seed <- 401

```

Read in metadata.
```{r echo = TRUE, eval = TRUE}

meta_dats <- read.xlsx(file.path(xls_path, "Mymon_RADlibraries_2015_2019_CombinedMetadata.xlsx"))
glimpse(meta_dats)

```

Read in the three HWE filtered Structure files: single-end, paired-end, merged, and the 2 versions of each (just HWE filtered, or additionally filtered by removing individuals and loci with a lot of missing data). These will be read in as genind objects. The number of loci per file is documented in HWE_filtering.Rmd.
```{r echo = TRUE, eval = FALSE}

single_end <- read.structure(file.path(file.path(res_path, "single_end"), "single_end_HWE_filter.str"), n.ind = 139, n.loc = 33726, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(single_end)

single_end_filt <- read.structure(file.path(file.path(res_path, "single_end"), "single_end_HWE_missingData_filters.str"), n.ind = 105, n.loc = 14381, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(single_end_filt)

paired_end <- read.structure(file.path(file.path(res_path, "paired_end"), "paired_end_HWE_filter.str"), n.ind = 143, n.loc = 6385, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(paired_end)

paired_end_filt <- read.structure(file.path(file.path(res_path, "paired_end"), "paired_end_HWE_missingData_filters.str"), n.ind = 83, n.loc = 2940, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(paired_end_filt)

merged <- read.structure(file.path(file.path(res_path, "merged"), "merged_HWE_filter.str"), n.ind = 282, n.loc = 869, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(merged)

merged_filt <- read.structure(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters.str"), n.ind = 216, n.loc = 561, col.lab = 1, col.pop = 2, row.marknames = 1, onerowperind = FALSE, ask = FALSE, NA.char = "-9999")
str(merged_filt)

# All loci are biallelic
# unique(single_end@loc.n.all)
# unique(paired_end@loc.n.all)
# unique(merged@loc.n.all)

```

Run DAPC per set of SNPs using sampling locations from sample metadata. Consider also doing this after filtering out individuals with a lot of missing data.

Single-end SNPs, HWE filter.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp1 <- sapply(1:nrow(single_end@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(single_end@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp1)
length(grp1)

# Replace the population slot with this info
# single_end@pop <- as.factor(grp)

# Change levels to be by country: Argentina, Spain, U.S.
grp1 <- factor(grp1, levels = c("ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
# Here selected the number of PCs that contained 60% of variation, and 2 discriminant functions
dapc1 <- dapc(single_end, pop = grp1, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 61 PCs retained
dapc1

# Make colors by country, hues by sampling location
levels(grp1)

# Argentina in grays, Spain in reds, US in golds
cols <- c("black", "gray5", "gray25", "gray50", "gray75", "firebrick", "red", "brown4", "coral2", "maroon", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(18, 5), rep(15, 6), rep(17, 4))

# dev.off()
scatter(dapc1, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

```

Single-end SNPs, HWE filter + missing data filters.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp1.2 <- sapply(1:nrow(single_end_filt@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(single_end_filt@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp1.2)
length(grp1.2)

# Replace the population slot with this info
# single_end@pop <- as.factor(grp)

# Change levels to be by country: Argentina, Spain, U.S.
levels(factor(grp1.2))

grp1.2 <- factor(grp1.2, levels = c("ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
# Here selected the number of PCs that contained 60% of variation, and 2 discriminant functions
dapc1.2 <- dapc(single_end_filt, pop = grp1.2, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 46 PCs retained
dapc1.2

# Make colors by country, hues by sampling location
levels(grp1.2)

# Argentina in grays, Spain in reds, US in golds
cols <- c("black", "gray5", "gray25", "gray50", "gray75", "firebrick", "red", "brown4", "coral2", "maroon", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(18, 5), rep(15, 6), rep(17, 4))

# dev.off()
scatter(dapc1.2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

```

Place both single end graphics in the same plot using ggplot.
```{r echo = TRUE, eval = FALSE}

data.frame(
  X = c(dapc1$ind.coord[, 1], dapc1.2$ind.coord[, 1]),
  Y = c(dapc1$ind.coord[, 2], dapc1.2$ind.coord[, 2])
  ) %>%
  dplyr::mutate(
    type = c(rep("HWE filter", nrow(dapc1$ind.coord)), rep("HWE + missing data filters", nrow(dapc1.2$ind.coord))),
    indiv = c(dimnames(dapc1$ind.coord)[[1]], dimnames(dapc1.2$ind.coord)[[1]]),
    site = c(as.character(grp1), as.character(grp1.2)),
    site = factor(site, levels = c("ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH")),
    type = factor(type, levels = c("HWE filter", "HWE + missing data filters"))
  ) %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = site)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ type) +
  scale_shape_manual(values = pchs) +
  scale_color_manual(values = alpha(cols, 0.75)) +
  scale_fill_manual(values = cols) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

```

Paired-end SNPs, HWE filter.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp2 <- sapply(1:nrow(paired_end@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(paired_end@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp2)
length(grp2)
unique(grp2)

# Change levels to be by country: Uruguay, Argentina, Spain, U.S.
levels(factor(grp2))

grp2 <- factor(grp2, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
dapc2 <- dapc(paired_end, pop = grp2, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 54 PCs retained
dapc2

# Make colors by country, hues by sampling location
levels(grp2)

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "maroon", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 4), rep(17, 4))

# dev.off()
scatter(dapc2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

# Zoom in
str(dapc2$ind.coord)
range(dapc2$ind.coord[, 1])
range(dapc2$ind.coord[, 2])

scatter(dapc2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE, ratio.pca = 1, xlim = c(-12, 12), ylim = c(-4, 4))
dev.off()

```

Paired-end SNPs, HWE + missing data filters.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp2.2 <- sapply(1:nrow(paired_end_filt@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(paired_end_filt@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp2.2)
length(grp2.2)
unique(grp2.2)

# Change levels to be by country: Uruguay, Argentina, Spain, U.S.
levels(factor(grp2.2))

grp2.2 <- factor(grp2.2, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
dapc2.2 <- dapc(paired_end_filt, pop = grp2.2, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 38 PCs retained
dapc2.2

# Make colors by country, hues by sampling location
levels(grp2.2)

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "maroon", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 9), rep(18, 5), rep(15, 4), rep(17, 4))

# dev.off()
scatter(dapc2.2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

# Zoom in
str(dapc2.2$ind.coord)
range(dapc2.2$ind.coord[, 1])
range(dapc2.2$ind.coord[, 2])

scatter(dapc2.2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE, ratio.pca = 1, xlim = c(-12, 12), ylim = c(-4, 4))
dev.off()

```

Place both paired-end graphics in the same plot using ggplot.
```{r echo = TRUE, eval = FALSE}

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "maroon", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 4), rep(17, 4))

data.frame(
  X = c(dapc2$ind.coord[, 1], dapc2.2$ind.coord[, 1]),
  Y = c(dapc2$ind.coord[, 2], dapc2.2$ind.coord[, 2])
  ) %>%
  dplyr::mutate(
    type = c(rep("HWE filter", nrow(dapc2$ind.coord)), rep("HWE + missing data filters", nrow(dapc2.2$ind.coord))),
    indiv = c(dimnames(dapc2$ind.coord)[[1]], dimnames(dapc2.2$ind.coord)[[1]]),
    site = c(as.character(grp2), as.character(grp2.2)),
    site = factor(site, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "GRCA", "CNCT", "FLOR", "ILLI", "WASH")),
    type = factor(type, levels = c("HWE filter", "HWE + missing data filters"))
  ) %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = site)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ type) +
  scale_shape_manual(values = pchs) +
  scale_color_manual(values = alpha(cols, 0.75)) +
  scale_fill_manual(values = cols) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

```

Note that some patterns change depending on the number of PCs retained (I didn't play around with the number of discriminant functions), but the overall overlap between most samples, including a lot of URY samples, with GRCA and ILLI falling out as most different, remains. These patterns really do seem due to missing data...note how much more similar the PE library patterns are to the SE library patterns when individual and loci with a lot of missing data are removed.

Merged SNPs, HWE filter.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp3 <- sapply(1:nrow(merged@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(merged@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp3)
length(grp3)
unique(grp3)

# Change levels to be by country: Uruguay, Argentina, Spain, U.S.
levels(factor(grp3))

grp3 <- factor(grp3, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
# dapc(merged, pop = grp)
dapc3 <- dapc(merged, pop = grp3, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 67 PCs retained
dapc3

# Make colors by country, hues by sampling location
levels(grp3)

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

dev.off()
scatter(dapc3, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

```

Merged SNPs, HWE + missing data filters.
```{r echo = TRUE, eval = FALSE}

# Get sample locations from metadata in the same order of samples in the genind object
grp3.2 <- sapply(1:nrow(merged_filt@tab), function(i){
  meta_dats$Site_Code[grep(paste("^", dimnames(merged_filt@tab)[[1]][i], "$", sep = ""), meta_dats$Sample_Name)]
})
head(grp3.2)
length(grp3.2)
unique(grp3.2)

# Change levels to be by country: Uruguay, Argentina, Spain, U.S.
levels(factor(grp3.2))

grp3.2 <- factor(grp3.2, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))

# If all PCs used, then no points will show up, just labels for sampling sites
# dapc(merged, pop = grp)
dapc3.2 <- dapc(merged_filt, pop = grp3.2, n.da = 2, pca.select = "percVar", perc.pca = 60)

# 51 PCs retained
dapc3.2

# Make colors by country, hues by sampling location
levels(grp3.2)

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

dev.off()
scatter(dapc3.2, scree.da = FALSE, bg = "white", col = cols, pch = pchs, cell = 0, cstar = 0, solid = 0.75, cex = 3, clab = 0, leg = TRUE)

```

Place both paired-end graphics in the same plot using ggplot.
```{r echo = TRUE, eval = FALSE}

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

data.frame(
  X = c(dapc3$ind.coord[, 1], dapc3.2$ind.coord[, 1]),
  Y = c(dapc3$ind.coord[, 2], dapc3.2$ind.coord[, 2])
  ) %>%
  dplyr::mutate(
    type = c(rep("HWE filter", nrow(dapc3$ind.coord)), rep("HWE + missing data filters", nrow(dapc3.2$ind.coord))),
    indiv = c(dimnames(dapc3$ind.coord)[[1]], dimnames(dapc3.2$ind.coord)[[1]]),
    site = c(as.character(grp3), as.character(grp3.2)),
    site = factor(site, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH")),
    type = factor(type, levels = c("HWE filter", "HWE + missing data filters"))
  ) %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = site)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ type) +
  scale_shape_manual(values = pchs) +
  scale_color_manual(values = alpha(cols, 0.75)) +
  scale_fill_manual(values = cols) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

```

There are some interesting differences in population genetic structure between the paired-end SNPs (greater genomic resolution, less population variation) and the merged SNPs (lower genomic resolution, more population variation). Patterns in the merged SNP datasets were robust to missing data.

Making a graphic across SNP datasets and filtering versions.
```{r echo = TRUE, eval = FALSE}

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

# A data frame for text labels
lab_df <- data.frame(X = rep(-30, 6), Y = rep(20, 6)) %>%
  dplyr::mutate(
    SNP_set = c(
      rep("Single-end", 2),
      rep("Paired-end", 2),
      rep("Merged", 2)
    ),
    filter_type = c(
      rep(c("HWE filter", "HWE + missing data filters"), 3)
    ),
    filter_lab = c(
      "139 indivs,\n 30085 SNPs", 
      "105 indivs,\n 13241 SNPs",
      "143 indivs,\n 6385 SNPs", 
      "83 indivs,\n 2940 SNPs",
      "282 indivs,\n 869 SNPs",
      "216 indivs,\n 561 SNPs"
    ),
    SNP_set = factor(SNP_set, levels = c("Single-end", "Paired-end", "Merged")),
    filter_type = factor(filter_type, levels = c("HWE filter", "HWE + missing data filters"))
  )
    
lab_df

all_snps_df <- data.frame(
  X = c(
    dapc1$ind.coord[, 1], 
    dapc1.2$ind.coord[, 1],
    dapc2$ind.coord[, 1], 
    dapc2.2$ind.coord[, 1],
    dapc3$ind.coord[, 1], 
    dapc3.2$ind.coord[, 1]
    ),
  Y = c(
    dapc1$ind.coord[, 2], 
    dapc1.2$ind.coord[, 2],
    dapc2$ind.coord[, 2], 
    dapc2.2$ind.coord[, 2],
    dapc3$ind.coord[, 2], 
    dapc3.2$ind.coord[, 2]
  )
) %>%
  dplyr::mutate(
    SNP_set = c(
      rep("Single-end", nrow(dapc1$ind.coord) + nrow(dapc1.2$ind.coord)),
      rep("Paired-end", nrow(dapc2$ind.coord) + nrow(dapc2.2$ind.coord)),
      rep("Merged", nrow(dapc3$ind.coord) + nrow(dapc3.2$ind.coord))
    ),
    filter_type = c(
      rep("HWE filter", nrow(dapc1$ind.coord)), 
      rep("HWE + missing data filters", nrow(dapc1.2$ind.coord)),
      rep("HWE filter", nrow(dapc2$ind.coord)), 
      rep("HWE + missing data filters", nrow(dapc2.2$ind.coord)),
      rep("HWE filter", nrow(dapc3$ind.coord)), 
      rep("HWE + missing data filters", nrow(dapc3.2$ind.coord))
      ),
    indiv = c(
      dimnames(dapc1$ind.coord)[[1]], 
      dimnames(dapc1.2$ind.coord)[[1]],
      dimnames(dapc2$ind.coord)[[1]], 
      dimnames(dapc2.2$ind.coord)[[1]],
      dimnames(dapc3$ind.coord)[[1]], 
      dimnames(dapc3.2$ind.coord)[[1]]
      ),
    site = c(
      as.character(grp1), 
      as.character(grp1.2),
      as.character(grp2), 
      as.character(grp2.2),
      as.character(grp3), 
      as.character(grp3.2)
      ),
    SNP_set = factor(SNP_set, levels = c("Single-end", "Paired-end", "Merged")),
    filter_type = factor(filter_type, levels = c("HWE filter", "HWE + missing data filters")),
    site = factor(site, levels = c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH"))
  ) 

# No zoom, numbers are labeled
ggplot(data = all_snps_df, aes(x = X, y = Y)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6, aes(fill = site, color = site, shape = site)) +
  facet_grid(rows = vars(SNP_set), cols = vars(filter_type), scales = "free", switch = "y") +
  geom_text(data = lab_df, aes(x = X, y = Y, label = filter_lab), color = "black") +
  scale_shape_manual(values = pchs) +
  scale_color_manual(values = alpha(cols, 0.75)) +
  scale_fill_manual(values = cols) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
   guides(color = guide_legend(ncol = 1), fill = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  scale_y_continuous(limits = c(-25, 25)) +
  theme_bw() +
  theme(
    axis.title = element_text(size = 14),
    strip.text = element_text(size = 14),
    legend.direction = "vertical"
  )

ggsave(file.path(gpath, "DAPC_HWE_missingData_filters_3SNPsets.jpeg"), width = 8.5, height = 11, dpi = 300)

# Zoom in, no labels
ggplot(data = all_snps_df, aes(x = X, y = Y)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6, aes(fill = site, color = site, shape = site)) +
  facet_grid(rows = vars(SNP_set), cols = vars(filter_type), scales = "free", switch = "y") +
  scale_shape_manual(values = pchs) +
  scale_color_manual(values = alpha(cols, 0.75)) +
  scale_fill_manual(values = cols) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  guides(color = guide_legend(ncol = 1), fill = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  scale_x_continuous(limits = c(-10, 10)) +
  scale_y_continuous(limits = c(-10, 10)) +
  theme_bw() +
  theme(
    axis.title = element_text(size = 14),
    strip.text = element_text(size = 14),
    legend.direction = "vertical"
  )

ggsave(file.path(gpath, "DAPC_HWE_missingData_filters_3SNPsets_zoomed.jpeg"), width = 8.5, height = 11, dpi = 300)

```

## Positive controls across sequencing types

Check out positive controls across sequencing types (e.g. samples used in both the SE and PE libraries).
```{r echo = TRUE, eval = FALSE}

all_sites <- c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH")

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

pos_cntrl_axYrs <- data.frame(
  X = c(dapc3$ind.coord[, 1], dapc3.2$ind.coord[, 1]),
  Y = c(dapc3$ind.coord[, 2], dapc3.2$ind.coord[, 2])
  ) %>%
  dplyr::mutate(
    type = c(rep("HWE filter", nrow(dapc3$ind.coord)), rep("HWE + missing data filters", nrow(dapc3.2$ind.coord))),
    indiv = c(dimnames(dapc3$ind.coord)[[1]], dimnames(dapc3.2$ind.coord)[[1]]),
    site = c(as.character(grp3), as.character(grp3.2)),
    site = factor(site, levels = all_sites),
    type = factor(type, levels = c("HWE filter", "HWE + missing data filters"))
  ) %>%
  # Join with metadata to get positive control status
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Sequencing_Type, Positive_Control_AcrossYears),
    by = c("indiv" = "Sample_Name")
  ) %>%
  # Filter by positive controls across years
  filter(Positive_Control_AcrossYears == "Y") %>%
  droplevels()
glimpse(pos_cntrl_axYrs)
nrow(pos_cntrl_axYrs)

# Get individual IDs from sample names
PE_cntrls <- pos_cntrl_axYrs %>%
  filter(Sequencing_Type == "Paired-end") %>%
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      paste(strsplit(indiv[i], split = "_")[[1]][3:5], collapse = "_")
    })
  ) %>%
  pull(id) %>%
  unique()

SE_cntrls <- pos_cntrl_axYrs %>%
  filter(Sequencing_Type == "Single-end") %>%
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      paste(strsplit(indiv[i], split = "_")[[1]][3:5], collapse = "_")
    })
  ) %>%
  pull(id) %>%
  unique()

head(PE_cntrls)
head(SE_cntrls)

length(PE_cntrls)
length(SE_cntrls)

# INV_WASH_WA2 was included in the PE_cntrls, but was dropped in early pre-processing due to very low coverage
# Looks good to proceed
SE_cntrls[which(!SE_cntrls %in% PE_cntrls)]

# Make aesthetics consistent with the full set of samples
tmp_sites <- levels(pos_cntrl_axYrs$site)
tmp_sites  

cols_cntrls <- cols[grep(paste(paste("^", tmp_sites, "$", sep = ""), collapse = "|"), all_sites)]

pchs_cntrls <- pchs[grep(paste(paste("^", tmp_sites, "$", sep = ""), collapse = "|"), all_sites)]

# Positive controls group together by sampling site
pos_cntrl_axYrs %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = site)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ type) +
  scale_shape_manual(values = pchs_cntrls) +
  scale_color_manual(values = alpha(cols_cntrls, 0.75)) +
  scale_fill_manual(values = cols_cntrls) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

# Making another graphic of the HWE and missing data filter dataset only, with one panel per sampling site, and labels for individuals to better assess how close positive controls are in genetic space across libraries
# Here, made shapes the sequencing type
shps <- c(21, 24)

# Note that in these two plots, some positive controls will only show up once because they were likly dropped in the stricter filter imposed

# With just symbols indicating sequencing type
pos_cntrl_axYrs %>%
  filter(type == "HWE + missing data filters") %>%
  droplevels() %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = Sequencing_Type)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ site, scales = "free") +
  scale_shape_manual(values = shps) +
  scale_color_manual(values = alpha(cols_cntrls, 0.75)) +
  scale_fill_manual(values = cols_cntrls) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()


# With just text indicating unique ID per site
pos_cntrl_axYrs %>%
  filter(type == "HWE + missing data filters") %>%
  # Get a unique ID that will be easier to compare across libraries than sample names
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      strsplit(indiv[i], split = "_")[[1]][5]
    })
  ) %>%
  droplevels() %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = Sequencing_Type)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_text(aes(label = id), color = "black", size = 4) +
  facet_wrap(~ site, scales = "free") +
  scale_shape_manual(values = shps) +
  scale_color_manual(values = alpha(cols_cntrls, 0.75)) +
  scale_fill_manual(values = cols_cntrls) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

```

Positive controls across sequencing types (e.g. individuals sequenced in both the 2015 single-end and 2019 paired-end libraries) occupied very similar areas of low-dimensional acoustic space within their respective sampling sites. Indeed, some individuals overlapped entirely from one sequencing type to another. In other cases, there was some separation between sequencing types in genetic space, which is likely due to differences in coverage leading to slightly different genotype calls for some individuals.

## Positive controls within libraries

For the paired-end libraries, some individuals were sequenced twice as positive controls on the same plate.
```{r echo = TRUE, eval = FALSE}

all_sites <- c("CHAC", "ELGE", "EMBR", "SEMI-3", "SOLE", "1135", "1145", "BAGU", "BCAR", "PEIX", "ALGA", "BAIR", "LURO", "BAZO", "ERIO", "BARC", "MADR", "MALL", "SEVI", "ZARA", "GRCA", "CNCT", "FLOR", "ILLI", "WASH")

# Uruguay in blues, Argentina in grays, Spain in reds, US in golds
cols <- c("navy", "blue", "royalblue", "dodgerblue", "cornflowerblue", "cyan", "turquoise", "darkcyan", "deepskyblue3", "lightskyblue", "black", "gray5", "gray25", "gray50", "gray75", "firebrick", "brown4", "red", "maroon", "coral2", "orangered", "gold4",  "gold2", "yellow2", "orange")

# Shapes by country
pchs <- c(rep(19, 10), rep(18, 5), rep(15, 6), rep(17, 4))

pos_cntrl_wtnPE <- data.frame(
  X = c(dapc3$ind.coord[, 1], dapc3.2$ind.coord[, 1]),
  Y = c(dapc3$ind.coord[, 2], dapc3.2$ind.coord[, 2])
  ) %>%
  dplyr::mutate(
    type = c(rep("HWE filter", nrow(dapc3$ind.coord)), rep("HWE + missing data filters", nrow(dapc3.2$ind.coord))),
    indiv = c(dimnames(dapc3$ind.coord)[[1]], dimnames(dapc3.2$ind.coord)[[1]]),
    site = c(as.character(grp3), as.character(grp3.2)),
    site = factor(site, levels = all_sites),
    type = factor(type, levels = c("HWE filter", "HWE + missing data filters"))
  ) %>%
  # Join with metadata to get positive control status
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Sequencing_Type, Positive_Control_2019),
    by = c("indiv" = "Sample_Name")
  ) %>%
  # Filter by positive controls in the PE libraries
  filter(Positive_Control_2019 == "Y") %>%
  droplevels()
glimpse(pos_cntrl_wtnPE)
nrow(pos_cntrl_wtnPE)

# Get individual IDs from sample names
PE_cntrls <- pos_cntrl_wtnPE %>%
  filter(Sequencing_Type == "Paired-end") %>%
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      paste(strsplit(indiv[i], split = "_")[[1]][3:5], collapse = "_")
    })
  ) %>%
  pull(id) %>%
  unique()

# 4 individuals included as positive controls, e.g. sequenced twice in the same library
PE_cntrls
length(PE_cntrls)

meta_dats %>%
  filter(Positive_Control_2019 == "Y") %>%
  pull(Sample_Name) %>%
  unique()

# Make aesthetics consistent with the full set of samples
tmp_sites <- levels(pos_cntrl_wtnPE$site)
tmp_sites  

cols_cntrls <- cols[grep(paste(paste("^", tmp_sites, "$", sep = ""), collapse = "|"), all_sites)]

pchs_cntrls <- pchs[grep(paste(paste("^", tmp_sites, "$", sep = ""), collapse = "|"), all_sites)]

# Positive controls group together by sampling site
pos_cntrl_wtnPE %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = site)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_point(size = 6) +
  facet_wrap(~ type) +
  scale_shape_manual(values = pchs_cntrls) +
  scale_color_manual(values = alpha(cols_cntrls, 0.75)) +
  scale_fill_manual(values = cols_cntrls) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

# Making another graphic with labels for individuals to better assess how close positive controls are in genetic space within the PE libraries

# With just text indicating unique ID per site
pos_cntrl_wtnPE %>%
  # filter(type == "HWE + missing data filters") %>%
  # Get a unique ID that will be easier to compare across libraries than sample names
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      strsplit(indiv[i], split = "_")[[1]][5]
    })
  ) %>%
  droplevels() %>%
  ggplot(aes(x = X, y = Y, fill = site, color = site, shape = Sequencing_Type)) +
  geom_hline(aes(yintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_vline(aes(xintercept = 0), linetype = "solid", color = "black", size = 0.25) +
  geom_text(aes(label = id), color = "black", size = 4) +
  facet_wrap(~ type, scales = "free") +
  scale_shape_manual(values = shps) +
  scale_color_manual(values = alpha(cols_cntrls, 0.75)) +
  scale_fill_manual(values = cols_cntrls) +
  xlab("Dimension 1") +
  ylab("Dimension 2") +
  theme_bw()

```

These positive controls look pretty good. The duplicated individuals fall within genetic space for their same sampling site, and are close to each other in genetic space. The second sample for L98 is missing in the right panel due to the stricter filter imposed after HWE on missing data per individual and locus. These visuals of positive controls served to validate the data quality of libraries, and the patterns we're seeing in genetic space for the single-end, paired-end, and merged SNP datasets. 

Proceed by using the HWE + missing data filtered version of the merged SNPs for making Bayescan input files to detect outlier loci, and make sure to remove positive controls. When positive controls were across sequencing types, remove the SE data when both SE and PE data are available for the same control (the PE data is higher quality due to removal of PCR duplicates). Take the SE data if only SE is available. For the positive controls sequenced twice within the PE libraries that are both present in the strictly filtered version of the SNP dataset, randomly select one sample per individual. 

# Remove duplicated positive controls for merged SNPs

Read in the HWE and missing data filtered merged SNPs.
```{r echo = TRUE, eval = TRUE}

struc <- read.table(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters.str"), skip = 1)
# glimpse(struc)
dim(struc) # 432 rows, 563 columns
head(struc[, 1]) # Sample names contained in the first column
head(struc[, 2]) # Population ID (single population) contained in second column
names(struc)

# Get the column headers

# Read the first line of metadata, these are the column names
col_nms <- readLines(file.path(file.path(res_path, "merged"), "merged_HWE_missingData_filters.str"), n = 1)
col_nms

# Split out the column headers into a vector, one element per header
mrkrs <- strsplit(strsplit(col_nms, split = "\t")[[1]][3], split = "[[:space:]]")[[1]]
head(mrkrs)
length(mrkrs)

# Get rid of the space that precedes the first locus column header
mrkrs <- mrkrs[-which(mrkrs == "")]
head(mrkrs)
length(mrkrs)

# Add the individual and population column headers that precede the locus column names
col_nms <- c("indiv", "pop", mrkrs)
head(col_nms)
length(col_nms)

# Add back the updated column names
# Looks good
names(struc) <- col_nms
glimpse(struc[, 1:10])

# Merge with the positive control columns from the metadata
struc2 <- struc %>%
  inner_join(
    meta_dats %>%
      dplyr::select(Sample_Name, Positive_Control_AcrossYears, Positive_Control_2019, Sequencing_Type),
    by = c("indiv" = "Sample_Name")
  ) %>%
  # Make a simpler unique ID per sample
  dplyr::mutate(
    indiv = as.character(indiv)
  ) %>%
  dplyr::mutate(
    id = sapply(1:nrow(.), function(i){
      paste(strsplit(indiv[i], split = "_")[[1]][3:5], collapse = "_")
    })
  ) %>%
  dplyr::select(indiv, pop, id, Positive_Control_AcrossYears, Positive_Control_2019, Sequencing_Type, all_of(mrkrs))

glimpse(struc2[, 1:10])

# Get positive controls across years (e.g. sequencing types) that should be dropped
ids2drop_se_pe <- struc2 %>%
  dplyr::select(indiv, pop, id, Positive_Control_AcrossYears) %>%
  filter(Positive_Control_AcrossYears == "Y") %>%
  group_by(id) %>%
  dplyr::summarise(
    n = length(id)
  ) %>%
  filter(n > 2) %>%
  pull(id) %>%
  unique()

# 40 positive controls across sequencing types are present more than twice in this data
# This means the SE samples for all of these IDs should be dropped
head(ids2drop_se_pe)
length(ids2drop_se_pe)

struc3 <- struc2 %>%
  dplyr::mutate(
    id_st = paste(id, Sequencing_Type, sep = "-")
  ) %>%
  filter(!id_st %in% paste(ids2drop_se_pe, "Single-end", sep = "-")) %>%
  droplevels()

glimpse(struc3[, 1:10])

# 40 IDs were dropped
nrow(struc2) - nrow(struc3)
length(unique(struc2$indiv)) - length(unique(struc3$indiv))

# Samples that were dropped, all single-end samples
unique(struc2$indiv)[which(!unique(struc2$indiv) %in% unique(struc3$indiv))]
ids2drop_se_pe

# Remove duplicated positive controls from the 2019 paired-end libraries
ids2drop <- struc3 %>%
  dplyr::select(indiv, pop, id, Positive_Control_2019) %>%
  filter(Positive_Control_2019 == "Y") %>%
  group_by(id) %>%
  dplyr::summarise(
    n = length(id)
  ) %>%
  filter(n > 2) %>%
  pull(id)

# Only 3 of the 4 2019 positive controls have more than one sample in this dataset
ids2drop

# Randomly select one sample to drop per positive control, the other will be kept
set.seed(seed)
i2drop <- struc3 %>%
  filter(id %in% ids2drop) %>%
  group_by(id) %>%
  dplyr::summarise(
    sample2drop = sample(indiv, 1, replace = FALSE)
  ) %>%
  pull(sample2drop)
i2drop

# Remove these positive controls
struc4 <- struc3 %>%
  filter(!indiv %in% i2drop) %>%
  droplevels() %>%
  dplyr::select(-c(id, id_st, Positive_Control_AcrossYears, Positive_Control_2019, Sequencing_Type))

dim(struc4)
nrow(struc4)/2 # 173 individuals remain
glimpse(struc4[, 1:10])

# Are all individuals represented twice as expected (diploid individuals get 2 rows in Structure format)
# Yes all individuals have 2 rows
struc4 %>% 
  group_by(indiv) %>%
  dplyr::summarise(n = length(indiv)) %>%
  pull(n) %>%
  unique()

# Write out the new version of this Structure object that has no duplicated positive controls
# Create a new Structure file, then proceed with the plan above
# Opened this file periodically in Vim to doublecheck structure
# Used :set list in Vim to see invisible characters
file_nm <- "merged_HWE_missingData_filters_noPosControlDups.str"

# Remove previous versions
file.remove(file.path(file.path(res_path, "merged"), file_nm))

# Initialize the Structure header, with the two tab symbols at the beginning and a new line symbol at the end
header <- c("\t", "\t", paste(paste(names(struc4)[-grep("^indiv$|^pop$", names(struc4))], collapse = " "), "\n", sep = ""))
# header

# Print the Structure header
cat(header, file = file.path(file.path(res_path, "merged"), file_nm), append = TRUE)

# Print each line of the filtered data frame above
invisible(pblapply(1:nrow(struc4), function(i){

  cat(c(struc4$indiv[i], "\t", struc4$pop[i], "\t", paste(paste(struc4[i, -grep("^indiv$|^pop$", names(struc4))], collapse = " "), "\n", sep = "")), file = file.path(file.path(res_path, "merged"), file_nm), append = TRUE)

}))

# Opened the file in Vim to check out results, looks good

```